packages/core/src/electron-main/electron-main-application.ts

```diff
@@ -0,0 +1,216 @@
+        const height = Math.floor(bounds.height * (2 / 3));
+        const width = Math.floor(bounds.width * (2 / 3));
+
+        const y = Math.floor(bounds.y + (bounds.height - height) / 2);
+        const x = Math.floor(bounds.x + (bounds.width - width) / 2);
+
+        const WINDOW_STATE = 'windowstate';
+        const windowState = this.electronStore.get(WINDOW_STATE, {
+            width, height, x, y
+        });
+
+        const windowOptions: electron.BrowserWindowConstructorOptions & {
+            isMaximized?: boolean,
+        } = {
+            show: false,
+            title: this.applicationName,
+            width: windowState.width,
+            height: windowState.height,
+            minWidth: 200,
+            minHeight: 120,
+            x: windowState.x,
+            y: windowState.y,
+            isMaximized: windowState.isMaximized
+        };
+
+        // Always hide the window, we will show the window when it is ready to be shown in any case.
+        const newWindow = new BrowserWindow(windowOptions);
+        if (windowOptions.isMaximized) {
+            newWindow.maximize();
+        }
+        newWindow.on('ready-to-show', () => newWindow.show());
+
+        // Prevent calls to "window.open" from opening an ElectronBrowser window,
+        // and rather open in the OS default web browser.
+        newWindow.webContents.on('new-window', (event, url) => {
+            event.preventDefault();
+            shell.openExternal(url);
+        });
+
+        // Save the window geometry state on every change
+        const saveWindowState = () => {
+            try {
+                // tslint:disable-next-line:no-shadowed-variable
+                let bounds: electron.Rectangle;
+                if (newWindow.isMaximized()) {
+                    bounds = this.electronStore.get(WINDOW_STATE, {});
+                } else {
+                    bounds = newWindow.getBounds();
+                }
+                this.electronStore.set(WINDOW_STATE, {
+                    isMaximized: newWindow.isMaximized(),
+                    width: bounds.width,
+                    height: bounds.height,
+                    x: bounds.x,
+                    y: bounds.y
+                });
+            } catch (e) {
+                console.error('Error while saving window state.', e);
+            }
+        };
+        // tslint:disable-next-line:no-any
+        let delayedSaveTimeout: any;
+        const saveWindowStateDelayed = () => {
+            if (delayedSaveTimeout) {
+                clearTimeout(delayedSaveTimeout);
+            }
+            delayedSaveTimeout = setTimeout(saveWindowState, 1000);
+        };
+        newWindow.on('close', saveWindowState);
+        newWindow.on('resize', saveWindowStateDelayed);
+        newWindow.on('move', saveWindowStateDelayed);
+
+        // Notify the renderer process on keyboard layout change
+        nativeKeymap.onDidChangeKeyboardLayout(() => {
```

Comment:
We want to refactor this code to move to its own service. Could it be done with `ElectronMainApplicationContribution.start` lifecycle callback?

Reply:
It should indeed, good idea